src/back/glsl_common.rs

```diff
@@ -0,0 +1,704 @@
+                    BinaryOperator::Multiply => "*",
+                    BinaryOperator::Divide => "/",
+                    BinaryOperator::Modulo => "%",
+                    BinaryOperator::Equal => "==",
+                    BinaryOperator::NotEqual => "!=",
+                    BinaryOperator::Less => "<",
+                    BinaryOperator::LessEqual => "<=",
+                    BinaryOperator::Greater => ">",
+                    BinaryOperator::GreaterEqual => ">=",
+                    BinaryOperator::And => "&",
+                    BinaryOperator::ExclusiveOr => "^",
+                    BinaryOperator::InclusiveOr => "|",
+                    BinaryOperator::LogicalAnd => "&&",
+                    BinaryOperator::LogicalOr => "||",
+                    BinaryOperator::ShiftLeftLogical => "<<",
+                    BinaryOperator::ShiftRightLogical => todo!(),
+                    BinaryOperator::ShiftRightArithmetic => ">>",
+                },
+                builder.expressions[*right].write_glsl(module, builder)?
+            ),
+            Expression::Intrinsic { fun, argument } => todo!(),
+            Expression::DotProduct(left, right) => format!(
+                "dot({},{})",
+                builder.expressions[*left].write_glsl(module, builder)?,
+                builder.expressions[*right].write_glsl(module, builder)?
+            ),
+            Expression::CrossProduct(left, right) => format!(
+                "cross({},{})",
+                builder.expressions[*left].write_glsl(module, builder)?,
+                builder.expressions[*right].write_glsl(module, builder)?
+            ),
+            Expression::Derivative { axis, expr } => format!(
+                "{}({})",
+                match axis {
+                    DerivativeAxis::X => "dFdx",
+                    DerivativeAxis::Y => "dFdy",
+                    _ => todo!(),
+                },
+                builder.expressions[*expr].write_glsl(module, builder)?
+            ),
+            Expression::Call { origin, arguments } => format!(
+                "{}({})",
+                match origin {
+                    FunctionOrigin::External(name) => name,
+                    FunctionOrigin::Local(handle) => builder.functions.get(&handle).unwrap(),
+                },
+                arguments
+                    .iter()
+                    .map(|arg| builder.expressions[*arg].write_glsl(module, builder))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .join(","),
+            ),
+        })
+    }
+}
+
+impl Constant {
+    pub(crate) fn write_glsl(
+        &self,
+        module: &Module,
+        builder: &StatementBuilder<'_>,
+    ) -> Result<String, Error> {
+        Ok(match &self.inner {
+            ConstantInner::Sint(int) => int.to_string(),
+            ConstantInner::Uint(int) => int.to_string(),
+            ConstantInner::Float(float) => float.to_string(),
```

Comment:
probably want to ensure here that there is at least one fraction digit, to not confuse GLSL